source/tests/utilities/tcp_echo_server.h

```diff
@@ -0,0 +1,164 @@
+#include <cstring>
+#include <memory>
+#include <string>
+#include <thread>
+#ifdef _WIN32
+#include <winsock2.h>
+#else
+#include <arpa/inet.h>
+#include <sys/socket.h>
+#include <netinet/in.h>
+#include <unistd.h>
+#endif
+
+#ifdef _WIN32
+typedef int socklen_t;
+#else
+typedef int SOCKET;
+#endif
+
+class TcpEchoSession : public std::enable_shared_from_this<TcpEchoSession> {
+public:
+    TcpEchoSession(SOCKET socket_fd)
+    : socket_fd_(socket_fd)
+    , stop_session_(false)
+    {
+    }
+
+    ~TcpEchoSession() {
+        stop_session_ = true;
+#ifdef _WIN32
+    closesocket(socket_fd_);
+#else
+    close(socket_fd_);
```

Comment:
This should probably call `stop()` so that logic is single-sourced. Also, the calls to close the socket should be indented.